python/ray/rllib/agents/sac/sac_policy_graph.py

```diff
@@ -0,0 +1,233 @@
+            output_shape=action_space.shape,
+            **policy_kwargs)
+
+        self.log_alpha = tf.get_variable(
+            'log_alpha',
+            dtype=tf.float32,
+            initializer=0.0)
+        self.alpha = tf.exp(self.log_alpha)
+
+        Q_type = self.config['Q']['type']
+        # TODO(hartikainen): implement twin q
+        assert Q_type == 'FeedforwardQ', Q_type
+        Q_kwargs = self.config['Q']['kwargs']
+        self.Q = feedforward_model(
+            input_shape=observation_space.shape,
+            output_size=1,
+            **Q_kwargs)
+        self.Q_target = tf.keras.models.clone_model(self.Q)
+
+    def _init_actor_loss(self):
+        actions = self.policy.actions([self._observations_ph])
+        log_pis = self.policy.log_pis([self._observations_ph], actions)
+
+        assert log_pis.shape.as_list() == [None, 1]
+
+        Q_log_targets = self.Q([self._observations_ph, actions])
+
+        # assert self._reparameterize
+
+        policy_kl_losses = self.alpha * log_pis - Q_log_targets
+
+        assert policy_kl_losses.shape.as_list() == [None, 1]
+
+        self.policy_loss = tf.reduce_mean(policy_kl_losses)
+
+    def _init_critic_loss(self):
+        next_actions = self.policy.actions([self._next_observations_ph])
+        next_log_pis = self.policy.log_pis(
+            [self._next_observations_ph], next_actions)
+
+        next_Q_values = self.Q([self._next_observations_ph, next_actions])
+        next_values = next_Q_values - self.alpha * next_log_pis
+
+        discount = self.config['gamma']
+
+        Q_targets = tf.stop_gradient(
+            self._rewards_ph
+            + discount
+            * (1.0 - tf.to_float(self._terminals_ph)) * next_values)
+
+        assert Q_targets.shape.as_list() == [None, 1]
+
+        Q_values = self.Q([self._observations_ph, self._actions_ph])
+        self.Q_loss = tf.losses.mean_squared_error(
+            labels=Q_targets, predictions=Q_values, weights=0.5)
+
+    def _init_entropy_loss(self):
+        target_entropy = (
+            -np.prod(self.action_space.shape)
+            if self.config['target_entropy'] == 'auto'
+            else self.config['target_entropy'])
+
+        assert isinstance(target_entropy, Number)
+
+        actions = self.policy.actions([self._observations_ph])
+        log_pis = self.policy.log_pis([self._observations_ph], actions)
+
+        self.entropy_loss = -tf.reduce_mean(
+            self.log_alpha * tf.stop_gradient(log_pis + target_entropy))
+
+    def _init_losses(self):
+        self._init_actor_loss()
+        self._init_critic_loss()
+        self._init_entropy_loss()
+
+        self.total_loss = (
+            self.policy_loss
+            + self.Q_loss
+            + self.entropy_loss)
+
+    @override(TFPolicyGraph)
+    def optimizer(self):
+        config = self.config['optimization']
+
+        entropy_optimizer = tf.train.AdamOptimizer(
+            learning_rate=config["entropy_lr"])
```

Comment:
Your ddpg policy graph uses single optimizer for both actor and critic. What's the reason for that? Or to put it other way, is there any reason to *not* use multiple optimizers for the losses? I would really like to have different optimizers for the losses, but the `TFPolicyGraph` needs to be changed to support that.

Reply:
Isn't having separate optimizers equivalent to using a single one and weighting the losses with coefficients?

Reply:
Yeah, I believe that's true. But if using weighting for the losses, you would have to consider the weights as hyperparameters, right? With separate optimizers we wouldn't have to worry about that.